include/wil/filesystem.h

```diff
@@ -1043,8 +1046,224 @@ namespace wil
+    // cases from the "create a new file" that has 4 variations represented in the create_file_create_behavior enum.
+
+    // https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilew?devlangs=cpp&f1url=%3FappId%3DDev16IDEF1%26l%3DEN-US%26k%3Dk(FILEAPI%252FCreateFileW)%3Bk(CreateFileW)%3Bk(DevLang-C%252B%252B)%3Bk(TargetOS-Windows)%26rd%3Dtrue
+
+    struct file_and_error_result
+    {
+        file_and_error_result(HANDLE file_handle, DWORD error) : file(file_handle), last_error(error)
+        {
+        }
+
+        wil::unique_hfile file;
+        DWORD last_error{};
+    };
+
+    /** Non-throwing open existing using OPEN_EXISTING.
+    ~~~
+    auto handle = wil::try_open_file(filePath.c_str());
+    ~~~
+    */
+    inline file_and_error_result try_open_file(PCWSTR path, DWORD dwDesiredAccess = FILE_READ_ACCESS,
+        DWORD dwShareMode = FILE_SHARE_READ,
+        bool inheritHandle = false,
+        DWORD dwFlagsAndAttributes = FILE_ATTRIBUTE_NORMAL) noexcept
+    {
+        SECURITY_ATTRIBUTES secAttributes{ sizeof(secAttributes) };
+        secAttributes.bInheritHandle = inheritHandle;
+        return { CreateFileW(path, dwDesiredAccess, dwShareMode, &secAttributes, OPEN_EXISTING, dwFlagsAndAttributes, nullptr),
+            ::GetLastError() };
+    }
+
+    /** open existing using OPEN_EXISTING, throws on error.
+    ~~~
+    auto handle = wil::open_file(filePath.c_str());
+    ~~~
+    */
+    inline wil::unique_hfile open_file(PCWSTR path, DWORD dwDesiredAccess = FILE_READ_ACCESS,
+        DWORD dwShareMode = FILE_SHARE_READ,
+        bool inheritHandle = false,
+        DWORD dwFlagsAndAttributes = FILE_ATTRIBUTE_NORMAL)
+    {
+        auto result = try_open_file(path, dwDesiredAccess, dwShareMode, inheritHandle, dwFlagsAndAttributes);
+        THROW_WIN32_IF(result.last_error, !result.file.is_valid());
+        return std::move(result.file);
+    }
+
+    /// @cond
+    namespace details
+    {
+        template<DWORD dwCreateDisposition>
+        file_and_error_result create_file(PCWSTR path, DWORD dwDesiredAccess,
+            DWORD dwShareMode,
+            LPSECURITY_ATTRIBUTES lpSecurityAttributes,
+            DWORD dwFlagsAndAttributes,
+            HANDLE hTemplateFile) noexcept
+        {
+            return { CreateFileW(
+                path, dwDesiredAccess, dwShareMode, lpSecurityAttributes, dwCreateDisposition, dwFlagsAndAttributes, hTemplateFile),
+                ::GetLastError() };
+        }
+    }
+    /// @endcond
+
+
+    /** create using CREATE_NEW, returns handle and error code.
+    ~~~
+    auto [handle, error = wil::try_create_new_file(filePath.c_str());
+    ~~~
+    */
+    inline file_and_error_result try_create_new_file(PCWSTR path,
+        DWORD dwDesiredAccess = FILE_READ_ACCESS | FILE_WRITE_ACCESS,
+        DWORD dwShareMode = FILE_SHARE_READ,
+        LPSECURITY_ATTRIBUTES lpSecurityAttributes = nullptr,
+        DWORD dwFlagsAndAttributes = FILE_ATTRIBUTE_NORMAL,

```

Comment:
I realize this is preserving argument order for `CreateFileW`, which might be good for reducing confusion, _but_ we have the opportunity to change it if we wanted to. E.g. I find myself specifying `dwFlagsAndAttributes` far more often than I find myself specifying `lpSecurityAttributes`.

Reply:
+1

Isn't it the case that if you want to specify an optional parameter then that means you must _also_ specify every optional parameter that comes before that on the argument list?

Reply:
> you must also specify every optional parameter that comes before that on the argument list

Correct

Reply:
I like the order that we were provided.
read/write mode and sharing options are the most commonly specified.
FILE_FLAG_BACKUPSEMANTICs is needed in the attributes for folders. so that is not as convenient as I'd like. but maybe that means we should add `wil::open_directory()`